src/lsm/forest_fuzz.zig

```diff
@@ -273,41 +355,81 @@ const Environment = struct {
-                    const lsm_account = env.get_account(id);
-
-                    // Compare result to model.
-                    const model_account = model.get(id);
-                    if (model_account == null) {
-                        assert(lsm_account == null);
-                    } else {
-                        assert(std.mem.eql(
-                            u8,
-                            std.mem.asBytes(&model_account.?),
-                            std.mem.asBytes(&lsm_account.?),
-                        ));
+            try env.apply_op(fuzz_op, &model);
+        }
+    }
+
+    fn apply_op(env: *Environment, fuzz_op: FuzzOp, model: anytype) !void {
+        switch (fuzz_op) {
+            .compact => |compact| {
+                env.compact(compact.op);
+                if (compact.checkpoint) {
+                    try model.checkpoint(compact.op);
+                    env.checkpoint(compact.op);
+                }
+            },
+            .put_account => |put| {
+                // The forest requires prefetch before put.
+                env.prefetch_account(put.account.id);
+                env.put_account(&put.account);
+                try model.put_account(&put.account, put.op);
+            },
+            .get_account => |id| {
+                // Get account from lsm.
+                env.prefetch_account(id);
+                const lsm_account = env.get_account(id);
+
+                // Compare result to model.
+                const model_account = model.get_account(id);
+                if (model_account == null) {
+                    assert(lsm_account == null);
+                } else {
+                    assert(std.mem.eql(
+                        u8,
+                        std.mem.asBytes(&model_account.?),
+                        std.mem.asBytes(&lsm_account.?),
+                    ));
+                }
+            },
+            .storage_reset => {
+                env.close();
+                env.deinit();
+                env.storage.reset();
+                try env.init(env.storage);
+
+                env.state = .superblock_open;
+                try env.open();
+
+                // TODO: currently this checks that everything added to the LSM after checkpoint
+                // resets to the last checkpoint on crash by looking through what's been added
+                // afterwards. This won't work if we add account removal to the fuzzer though.
+                const log_size = model.log.readableLength();
+                var log_left = log_size;
+                while (log_left > 0) : (log_left -= 1) {
+                    const entry = model.log.peekItem(log_size - log_left);
+                    const id = entry.account.id;
+                    if (model.checkpointed.get(id)) |checkpointed_account| {
+                        env.prefetch_account(id);
+                        if (env.get_account(id)) |lsm_account| {
+                            assert(std.mem.eql(
```

Comment:
Should this also assert that the lsm_account != entry.account?

Reply:
If `lsm_account` == `entry.account` then `lsm_account` != `checkpointed_account`, unless `entry.account` == `checkpointed_account`, in which case it's a false positive anyway. So I don't think it buys anything.